### PR TITLE
Update prune's behaviour for namespaces

### LINKED ISSUE #110905: kubectl apply: prune should ignore no-namespaced resource if -n is not empty
seems I have  the same problem. save this to a file named `a.yaml`
```
apiVersion: v1
kind: Service
metadata:
  annotations:
    environment: sandbox
    provider: kustomize
  labels:
    app: nginx
  name: nginx
  namespace: sandbox
spec:
  ports:
  - port: 80
  selector:
    app: nginx
  type: NodePort
```

And run `kubectl apply --prune --dry-run=client --all -n sandbox -f a.yaml`, the results
```
service/nginx created (dry run)
namespace/cattle-system pruned (dry run)
```

I have no idea why it want to prune the `namespace/cattle-system`.

_Originally posted by @wd in https://github.com/kubernetes/kubernetes/issues/66430#issuecomment-708922169_

### PR DESCRIPTION
**Pull request by Affan-7**
#### What type of PR is this?

/kind bug
/kind deprecation

#### What this PR does / why we need it:

This PR changes the behaviour of `--prune` flag of `kubectl apply`.

Previously if you use `--prune` with a namespace flag like ` -n sandbox` it also prunes other non-namespaced items as described in #110905.

>Run `kubectl apply --prune --dry-run=server --all -n sandbox -f a.yaml`, this results in
>service/nginx created (dry run)
>namespace/cattle-system pruned (dry run)

After this PR, the `--prune` flag no longer prunes non-namespaced items if used with a namesapce falg like `-n sandbox`.

>Run `kubectl apply --prune --dry-run=server --all -n sandbox -f a.yaml`, this results in
>service/nginx created (dry run)

(It still prunes the non-namespaced items if namespace is not sepcifed )

>Run `kubectl apply --prune --dry-run=server --all -f b.yaml`, this results in
>service/nginx created (dry run)
>namespace/cattle-system pruned (dry run)

#### Which issue(s) this PR fixes:
Fixes #110905

#### Does this PR introduce a user-facing change?

```release-note
Action Required: kubectl apply will no longer prune non-namespaced resources by default when used with the --namespace flag. To preserve the current behaviour, list the resources you want to target explicitly in the --prune-allowlist flag.
```

### CODE REVIEW

**Comment by pacoxu** [comment]
`pull-kubernetes-unit` failure should be fixed. 


> {Failed;Failed;Failed;Failed;Failed;  === RUN   TestApplyPruneObjectsWithAllowlist/prune_without_namespace_and_allowlist_should_delete_resources_that_are_not_in_the_specified_file
>     apply_test.go:1043: expected output to contain "namespace/test-apply pruned", but it did not. Actual Output:
>         replicationcontroller/test-rc unchanged
>         configmap/test-cm pruned
>         replicationcontroller/test-rc2 pruned
>     --- FAIL: TestApplyPruneObjectsWithAllowlist/prune_without_namespace_and_allowlist_should_delete_resources_that_are_not_in_the_specified_file (0.03s)
>

File: staging/src/k8s.io/kubectl/pkg/cmd/apply/apply_test.go [hunk 1]
```diff
@@ -905,17 +905,14 @@ func TestApplyPruneObjectsWithAllowlist(t *testing.T) {
 				"namespace/test-apply pruned",
```
Comment (pacoxu) [inline staging/src/k8s.io/kubectl/pkg/cmd/apply/apply_test.go:905]: This test case should be updated as well.

**Comment by pacoxu** [comment]
/priority important-soon
/triage accepted
As there is an **action required** release note here, we should take care of this before merging.

File: staging/src/k8s.io/kubectl/pkg/cmd/apply/apply_test.go (hunk 1 above)
Comment (Affan-7) [reply staging/src/k8s.io/kubectl/pkg/cmd/apply/apply_test.go:905]: I think it should prune the non-namespaced items when a namespace and a allow list is not specified.

It's interesting that this test case is failing, but this test under the [prune_test.go](https://github.com/kubernetes/kubernetes/blob/cd04f47e8b287068e8bf0bfae6596b2606d54929/staging/src/k8s.io/kubectl/pkg/util/prune/prune_test.go#L58C1-L66C4) is passing.

```
{
	mapper:             &testRESTMapper{},
	pr:                 []Resource{},
	namespaceSpecified: false,
	expectedns:         14,
	expectednns:        2,
	expectederr:        nil,
},
```

Comment (pacoxu) [reply staging/src/k8s.io/kubectl/pkg/cmd/apply/apply_test.go:905]: > It's interesting that this test case is failing, but this test under the [prune_test.go](https://github.com/kubernetes/kubernetes/blob/cd04f47e8b287068e8bf0bfae6596b2606d54929/staging/src/k8s.io/kubectl/pkg/util/prune/prune_test.go#L58C1-L66C4) is passing.

This case should pass. 

- If the namespace is not specified, default 14 namespace resources and 2 none-namespaced resources will be pruned.
- If the namespace is specified, only the default 14 namespace resources will be pruned.

Comment (Affan-7) [reply staging/src/k8s.io/kubectl/pkg/cmd/apply/apply_test.go:905]: > > It's interesting that this test case is failing, but this test under the [prune_test.go](https://github.com/kubernetes/kubernetes/blob/cd04f47e8b287068e8bf0bfae6596b2606d54929/staging/src/k8s.io/kubectl/pkg/util/prune/prune_test.go#L58C1-L66C4) is passing.
> 
> This case should pass.
> 
> * If the namespace is not specified, default 14 namespace resources and 2 none-namespaced resources will be pruned.
> * If the namespace is specified, only the default 14 namespace resources will be pruned.

Yes that's true, but this test under the [apply_test.go](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/119687/pull-kubernetes-unit/1697893585954803712) is failing.

```
=== FAIL: vendor/k8s.io/kubectl/pkg/cmd/apply TestApplyPruneObjectsWithAllowlist/prune_without_namespace_and_allowlist_should_delete_resources_that_are_not_in_the_specified_file (0.03s)
    apply_test.go:1040: expected output to contain "namespace/test-apply pruned", but it did not. Actual Output:
        replicationcontroller/test-rc unchanged
        configmap/test-cm pruned
        replicationcontroller/test-rc2 pruned
```

It should have pruned the namesapce.

Comment (pacoxu) [reply staging/src/k8s.io/kubectl/pkg/cmd/apply/apply_test.go:905]: Sorry, after investigation, I think we have made a mistake here.

https://github.com/kubernetes/kubernetes/blob/3c4a813d9c475909b50d1ea510ae5f34b5c4b847/staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go#L462-L463

- the namespace will default to `default` if not set.

https://github.com/kubernetes/kubernetes/blob/3c4a813d9c475909b50d1ea510ae5f34b5c4b847/staging/src/k8s.io/kubectl/pkg/cmd/apply/prune.go#L74-L75

namespaceSpecified `o.Namespace != ""` will always be true. 

The warning message will be there for everyone(no matter users specified namespace for the `kubectl apply --prune`) if I understand correctly.

Comment (pacoxu) [reply staging/src/k8s.io/kubectl/pkg/cmd/apply/apply_test.go:905]: /hold

Comment (Affan-7) [reply staging/src/k8s.io/kubectl/pkg/cmd/apply/apply_test.go:905]: I wonder what should we do now?

Comment (mashetty330) [reply staging/src/k8s.io/kubectl/pkg/cmd/apply/apply_test.go:905]: @pacoxu I was referring to the original issue and I ended up in this PR. where does the validation for `the namespace will default to default if not set.` happens. I dont see it in DefaultNamespace() IIUC. Could you please provide the pointer to that

**Comment by Affan-7** [comment]
Hi @pacoxu, can you please provide some suggestions on how to fix this failing unit test?

**Comment by brianpursley** [comment]
Is anything else needed for this PR?

I took a look and it looks OK to me, but I'd like to see what @seans3 or @soltysh thinks about it since they are more familiar than I am with prune.

It could cause a problem if someone relies on the old behavior, but it has gone through the deprecation period with a warning, so it should be ok(?).

**Comment by Affan-7** [comment]
I don't think that this PR is producing much impact, closing this.

**Pull request closed without merging**